### PR TITLE
Flatten data structures in geometry

### PR DESCRIPTION
**Pull request by SarahRo**
Replace `std::array<std::array<double, 3>, 2>` with `std::array<double, 6>`, and `std::vector<std::array<int, 2>>` with `std::vector<int>`.

Avoid copying data in some places where it was unnecessary.

### CODE REVIEW

**Review by jorgensd — CHANGES_REQUESTED** [review]

**Review by jorgensd — APPROVED** [review]

File: cpp/dolfinx/geometry/BoundingBoxTree.cpp [hunk 1]
```diff
@@ -34,8 +34,8 @@ std::vector<std::int32_t> range(const mesh::Mesh& mesh, int tdim)
 }
 //-----------------------------------------------------------------------------
 // Compute bounding box of mesh entity
```
Comment (IgorBaratta) [inline cpp/dolfinx/geometry/BoundingBoxTree.cpp:36]: The data layout in `std::array<std::array<double, 3>, 2>` is quite evident, but not in `std::array<double, 6>`.
I understand that it uses `XYZXYZ` but had to look at the function internals.

Comment (jorgensd) [inline cpp/dolfinx/geometry/BoundingBoxTree.cpp:36]: Happy with the following
```suggestion
// Compute bounding box of mesh entity. The bounding box is defined by (lower left corner, top right corner).
// Storage flattened row-major
```
@IgorBaratta

File: cpp/dolfinx/geometry/BoundingBoxTree.cpp [hunk 2]
```diff
@@ -45,129 +45,114 @@ compute_bbox_of_entity(const mesh::Mesh& mesh, int dim, std::int32_t index)
...
+      b0[j] = std::min(b0[j], xg[3 * local_vertex + j]);
+      b1[j] = std::max(b1[j], xg[3 * local_vertex + j]);
     }
   }
 
   return b;
 }
 //-----------------------------------------------------------------------------
 // Compute bounding box of bounding boxes
-std::array<std::array<double, 3>, 2> compute_bbox_of_bboxes(
```
Comment (IgorBaratta) [inline cpp/dolfinx/geometry/BoundingBoxTree.cpp:67]: same as above

File: cpp/dolfinx/geometry/BoundingBoxTree.h [hunk 3]
```diff
@@ -103,6 +109,19 @@ class BoundingBoxTree
...
   }
 
+  /// Get bounding box child nodes
+  ///
+  /// @param[in] node The bounding box node index
+  /// @return The indices of the two child nodes. If @p node is a leaf
+  /// nodes, then the values in the returned array are equal and
+  /// correspond to the index of the entity that the leaf node bounds,
+  /// e.g. the index of the cell that it bounds.
+  std::span<const int, 2> bbox_span(std::size_t node) const
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/BoundingBoxTree.h:119]: This should return a `std::array<int, 2>`. No need for a span.

File: cpp/dolfinx/geometry/BoundingBoxTree.h [hunk 4]
```diff
@@ -70,9 +70,15 @@ class BoundingBoxTree
...
-  std::array<std::array<double, 3>, 2> get_bbox(std::size_t node) const;
+  /// @return A copy of bounding box (lower_corner, upper_corner). Shape (2,3).
+  /// Flattened row-major.
+  std::array<double, 6> copy_bbox(std::size_t node) const;
+
+  /// Return bounding box coordinates for a given node in the tree
+  /// @param[in] node The bounding box node index
+  /// @return The bounding box (lower_corner, upper_corner) as a subspan of the
+  /// bounding box coordinates. Shape (2,3). Flattened row-major.
+  std::span<const double, 6> get_bbox(std::size_t node) const;
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/BoundingBoxTree.h:81]: I think this should be removed, and `copy_bbox` renamed to `get_bbox` unless profiling shows a significant performance benefit. Returning a copy is better than a cont reference/pointer if there isn't a compelling performance reason.

File: cpp/dolfinx/geometry/utils.cpp [hunk 5]
```diff
@@ -29,16 +29,19 @@ constexpr bool is_leaf(const std::array<int, 2>& bbox)
...
 /// of the bounding box, b(0,i) <= x[i] <= b(1,i) for i = 0, 1, 2
-constexpr bool point_in_bbox(const std::array<std::array<double, 3>, 2>& b,
-                             const std::array<double, 3>& x)
+constexpr bool point_in_bbox(std::span<const double, 6> b,
+                             std::span<const double, 3> x)
 {
+  assert(b.size() == 6);
   constexpr double rtol = 1e-14;
   bool in = true;
+  auto b0 = b.subspan(0, 3);
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.cpp:38]: Does this syntax make the span size a compile-time constant?

There are examples at https://en.cppreference.com/w/cpp/container/span/dynamic_extent on how to check.

**Review by garth-wells — COMMENTED** [review]
Are the changes from `std::array<X, 6>` to  `std::span<const X, 6>` in interfaces really needed? The cost of a copy should be more-or-less trivial. Is there some performance data?

File: cpp/dolfinx/geometry/utils.cpp (hunk 5 above)
Comment (jorgensd) [reply cpp/dolfinx/geometry/utils.cpp:38]: We should use 
```c++
b.subspan<0,3>
```
As I tested it with:
```c++
td::vector<double> vec(10);
  std::span<double> sp(vec);
  print("span", sp.extent);
  constexpr std::size_t start = 0;
  constexpr std::size_t end = 3;

  auto ssp = sp.subspan(start, end);
  print("subspan", ssp.extent);
  std::span<double, 3> sssp = sp.subspan<0, 3>();
  print("constexpr subspan", sssp.extent);
```
yielding
```bash
span, dynamic extent
subspan, dynamic extent
constexpr subspan, static extent = 3
```
See https://en.cppreference.com/w/cpp/container/span/subspan for signature details. I'll fix this later today or tomorrow.

File: cpp/dolfinx/geometry/BoundingBoxTree.cpp [hunk 6]
```diff
@@ -45,129 +46,114 @@ compute_bbox_of_entity(const mesh::Mesh& mesh, int dim, std::int32_t index)
...
-      b[1][j] = std::max(b[1][j], xg[3 * local_vertex + j]);
+      b0[j] = std::min(b0[j], xg[3 * local_vertex + j]);
+      b1[j] = std::max(b1[j], xg[3 * local_vertex + j]);
     }
   }
 
   return b;
 }
 //-----------------------------------------------------------------------------
 // Compute bounding box of bounding boxes
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/BoundingBoxTree.cpp:70]:
```suggestion
// Compute bounding box of bounding boxes. Each bounding box is defined as a tuple (corners, entity_index).
// The corners of the bounding box is flattened row-major as (lower left corner, top right corner).
```

File: cpp/dolfinx/geometry/BoundingBoxTree.cpp [hunk 7]
```diff
@@ -45,129 +45,114 @@ compute_bbox_of_entity(const mesh::Mesh& mesh, int dim, std::int32_t index)
...
+                                 bboxes, bbox_coordinates);
 
     // Store bounding box data. Note that root box will be added last.
-    bboxes.push_back(bbox);
-    bbox_coordinates.insert(bbox_coordinates.end(), b0.begin(), b0.end());
-    bbox_coordinates.insert(bbox_coordinates.end(), b1.begin(), b1.end());
-    return bboxes.size() - 1;
+    bboxes.push_back(bbox0);
+    bboxes.push_back(bbox1);
+    common::impl::copy_N<6>(b.begin(), std::back_inserter(bbox_coordinates));
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply cpp/dolfinx/geometry/BoundingBoxTree.cpp]: I've changed it to use `std::copy_n` which I guess should not be significantly different from `std::copy`

**Comment by jorgensd** [comment]
This PR
```bash
Build tree (mus): 8889 11042 9456
Coll tree (mus): 555966 580818  566159
Build tree (mus): 8869 10062 9247
Coll tree (mus): 559656 581611  568570
Build tree (mus): 10099 11832 10811
Coll tree (mus): 621510 720529  663874
Build tree (mus): 9307 11173 9905
Coll tree (mus): 574994 626194  597880
```
Main
```bash
Build tree (mus): 9111 12733 10232
Coll tree (mus): 829225 955935  864322
Build tree (mus): 9145 11842 9804
Coll tree (mus): 830766 866580  840812
Build tree (mus): 9040 10724 9700
Coll tree (mus): 830286 906452  851835
```

File: cpp/dolfinx/geometry/BoundingBoxTree.h (hunk 4 above)
Comment (jorgensd) [reply cpp/dolfinx/geometry/BoundingBoxTree.h:80]: It does not seem like there is a benefit performance wise. These are the timings with `get_bbox`:
```bash
Build tree (mus): 8812 11677 9573
Coll tree (mus): 553568 576671  561480
Build tree (mus): 8945 11141 9416
Coll tree (mus): 554286 572038  562611
Build tree (mus): 9071 12135 9514
Coll tree (mus): 555176 565760  558456
Build tree (mus): 9360 13424 10399
Coll tree (mus): 567018 748228  648080
```
(same code as in https://github.com/FEniCS/dolfinx/pull/2313#issuecomment-1239332493)

**Comment by jorgensd** [comment]
@garth-wells I've now considered all your comments.
I've asked @IgorBaratta to verify the perfomance results on CSD3 (or a bigger system than a laptop).

**Comment by jorgensd** [comment]
Superseeded by #2359

**Pull request closed without merging**